_ramblings/2018-07-11-building-a-shell-part-1.md

```diff
@@ -0,0 +1,203 @@
+
+```rb
+# Print a prompt, then read a line from STDIN.
+def read
+  print('$ ')
+
+  # Try to read a line, exiting if there's no more lines.
+  line = STDIN.gets
+  if line
+    line.chomp.split(' ')
+  else
+    exit
+  end
+end
+```
+
+The following implementation of `evaluate` is a copout, because we just pass
+the input through to Ruby's implementation of [`system`][system] which calls
+`/bin/sh` under the hood. It does the correct thing however, and is enough to
+get us off the ground running.
+
+```rb
+# Run the given command in a subprocess.
+def evaluate(argv)
+  success = system(*argv)
+  unless success
+    puts("unknown command '#{argv[0]}'")
+  end
+end
+```
+
+Now that we've defined both `read` and `evaluate`, the full blown REPL is a
+simple one liner.
+
+```rb
+# The glorious REPL itself.
+loop { evaluate(read) }
+```
+
+Running this tiny Ruby program will give you a semi-functional shell. Take a
+moment and try it out for yourself! There are however, a few critical features
+missing, for example you'll notice `cd` doesn't work. This is where builtins
+come into play.
+
+### Extending Our Shell (a little)
+
+Builtins are commands that the shell executes directly, without calling out
+to a subprocess. This is exactly what we need for `cd`, since we don't want to
+change the working directory of a sub-process, we want to update the shell's
+working directory itself.
+
+We'll define a hash of `'command' => lambda { |args| ... }` to serve as a
+lookup table for the builtins.
+
+```rb
+# Commands to be executed directly by this shell.
+BUILTINS = {}
+```
+
+Next we need to update our `evaluate` function to handle our special builtin
+commands.
+
+```rb
+# Run the given command in a subprocess.
+def evaluate(argv)
+  if BUILTINS[argv[0]]
+    BUILTINS[argv[0]].call(argv[1..-1])
+  else
+    success = system(*argv)
+    unless success
+      puts("unknown command '#{argv[0]}'")
+    end
+  end
+end
+```
+
+Finally, we can start defining the needed builtins. We'll start with the two
+most needed commands, `cd` and `exit`, however there are many others which
+could be defined at this point. Mainly `.` (or `source`), `alias` and `exec`.
+I'll leave those implementations as exercises for the reader.
+
+```rb
+# The builtin `cd` for changing the shell's working directory.
+BUILTINS['cd'] = lambda do |args|
+  # Change to the home directory by default.
+  args << ENV['HOME'] if args.empty?
+
+  # Read the destination path, doing very basic path expansion.
+  dest = args.pop.gsub(/~/, ENV['HOME'])
+
+  # Try to change this shell's working directory.
+  begin
+    Dir.chdir(dest)
+  rescue Exception
+    puts("no such directory: #{dest}")
+  end
+end
+```
+
+```rb
+# The builtin to exit the shell.
+BUILTINS['exit'] = lambda do |args|
+  # Exit with a status of 0 by default.
+  args << 0 if args.empty?
+
+  # Exit the shell.
+  exit args.pop.to_i
+end
+```
+
+There's another small issue the observant reader may have noticed. If you enter
```

Comment:
A bit patronising again, I find it very tricky to address the reader.

Reply:
Again, style. And a nice reward to a reader who actually built the shell as she followed along, since I'd imagine they'd try to quit at some point with <kbd>ctrl-c</kbd>.